### PR TITLE
wp_update_post overwrites post_date when updating post_status

### PR DESCRIPTION
**Pull request by SirLouen**
For this reviewed patch I've considered these elements:

1. I've taken as the fix:  https://core.trac.wordpress.org/attachment/ticket/62468/fix-wp-update-post.patch
It was unnecesary to preserving date of `publish` and `private` on `draft/pending` posts. Only `future` is relevant in this scenario

2. For the unit tests I've [taken this one](https://github.com/WordPress/wordpress-develop/pull/8118.diff) by @sukhendu2002, removing all the extra AI fluff. The data provider feels a bit overkill but I think it will don't do any harm.

3. Finally, given that here is a good opportunity to open a specific file to cover `wp_update_posts` I've brought all the tests covering this function and merged them all into this file for Unit Tests organizational purposes.

Trac ticket: https://core.trac.wordpress.org/ticket/62468

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/post/wpInsertPost.php [hunk 1]
```diff
@@ -634,23 +634,6 @@ public function test_get_post_ancestors_within_loop() {
 		$this->assertSame( array( $parent_id ), get_post_ancestors( 0 ) );
 	}
 
-	/**
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpInsertPost.php:609]: I suggest keeping this here for the `wp_insert_post()` assertions, removing the `wp_update_post()` assertions in favor of the new file.

While not strictly related to this ticket, it will be handy for upcoming changes to phpunit https://core.trac.wordpress.org/ticket/62004#Problem4:coversattributesareonlysupportedatclasslevelx2

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	/**
+	 * Test updating a post and preserving post_date when changing to future status
+	 *
+	 * @ticket 62468
+	 *
+	 * @dataProvider data_post_status_transitions
+	 *
+	 * @param string $initial_status Initial post status.
+	 */
+	public function test_update_post_preserves_date( $initial_status ) {
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php]:
```suggestion
	public function test_update_post_preserves_date_for_future_posts( $initial_status ) {
```

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+		$this->assertSame( $future_date, $updated_post->post_date );
+	}
+
+	/**
+	 * Data provider for test_update_post_preserves_date
+	 *
+	 * @return array[] Test parameters
+	 */
+	public function data_post_status_transitions() {
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php]: This ought to have the same name as the test with the `test_` prefix removed and replaced with `data_`.

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+		$role = get_role( 'administrator' );
+		$role->add_cap( 'publish_mapped_meta_caps' );
+		$role->add_cap( 'publish_unmapped_meta_caps' );
+	}
+
+	/**
+	 * Tear down after class.
+	 */
+	public static function tear_down_after_class() {
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php]:
```suggestion
	public static function wpTearDownAfterClass() {
```

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+		$role = get_role( 'administrator' );
+		$role->remove_cap( 'publish_mapped_meta_caps' );
+		$role->remove_cap( 'publish_unmapped_meta_caps' );
+
+		parent::tear_down_after_class();
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php]:
```suggestion
```

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+			'ID'          => $post_id,
+			'post_status' => 'future',
+			'post_date'   => $future_date,
+		);
+
+		wp_update_post( $update_data );
+		$updated_post = get_post( $post_id );
+
+		$this->assertSame( $future_date, $updated_post->post_date );
+	}
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php:232]: Additional test cases

* unpublished statuses to future with a date in the past (I think now WP updates the date and replaces the future status to publish. This seems correct.)
* future to publish does update the date
*

**Review by peterwilsoncc — COMMENTED** [review]
I've added a few notes inline.

File: tests/phpunit/tests/post/wpUpdatePost.php (hunk 4 above)
Comment (SirLouen) [reply tests/phpunit/tests/post/wpUpdatePost.php]: Hello @peterwilsoncc, I copied this from WpInsertPost.php. It's called like this. There is no parent called `wpTearDownAfterClass`

PS: According to the WordPress CS, it seems that all classes in tests have been named wrongly with the lowercase `wp`

File: tests/phpunit/tests/post/wpUpdatePost.php (hunk 6 above)
Comment (SirLouen) [reply tests/phpunit/tests/post/wpUpdatePost.php:232]: I've opened a [new ticket #63911](https://core.trac.wordpress.org/ticket/63911#ticket) to add this cases, which are not directly related to this particular ticket, but improve coverage. I think this ticket is bringing too many changes with the bug + the merge of tests into a single file, so we can have things a little bit separated and provide more context individually. 
Once we merge this ticket, I will submit a PR with these two new tests in that ticket.

File: tests/phpunit/tests/post/wpUpdatePost.php (hunk 4 above)
Comment (peterwilsoncc) [reply tests/phpunit/tests/post/wpUpdatePost.php]: See https://github.com/WordPress/wordpress-develop/blob/6ae67d9bf64df5cea4be9468163c63f344e874fa/tests/phpunit/includes/abstract-testcase.php#L87-L89

It behaves the same way as `wpSetUpBeforeClass()` in that it doesn't require a call to the parent. If you search the code base the use of `wpTearDownAfterClass` is more common that.

File: tests/phpunit/tests/post/wpUpdatePost.php (hunk 6 above)
Comment (peterwilsoncc) [reply tests/phpunit/tests/post/wpUpdatePost.php:232]: I think these are related to this issue as the PR changes how `wp_update_post()` handles the date when dealing with the `future` status. 

This pull request needs to make sure that the status check doesn't affect other transitions.

Comment (SirLouen) [reply tests/phpunit/tests/post/wpUpdatePost.php:232]: I'm going to take advantage of the dataProvider to add all these scenarios.

File: tests/phpunit/tests/post/wpUpdatePost.php (hunk 4 above)
Comment (SirLouen) [reply tests/phpunit/tests/post/wpUpdatePost.php]: Interesting thing. Updated

**Comment by SirLouen** [comment]
@peterwilsoncc as a rule of thumb, in a bug, I don't like to add tests that cannot be proven with and without the patch (like in TDD). But I've taken advantage of the `dataProvider` to add a number of cases, including those you mentioned (some are implicit, for example, a publish to future in the past, ending in publish as expected).

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+		);
+
+		kses_remove_filters();
+
+		$post = get_post( $post->ID );
+		$this->assertSame( 'Test', $post->post_title );
+	}
+
+	/**
+	 * Test updating a post and preserving post_date when changing to future status
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/post/wpUpdatePost.php]:
```suggestion
	 * Test updating a post and preserving post_date when changing to future status.
```

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+		wp_update_post( $update_data );
+		$updated_post = get_post( $post_id );
+
+		$this->assertSame( $future_date, $updated_post->post_date );
+		$this->assertSame( $expected_status, $updated_post->post_status );
+	}
+
+	/**
+	 * Data provider for test_update_post_preserves_date_for_future_posts
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/post/wpUpdatePost.php]:
```suggestion
	 * Data provider for test_update_post_preserves_date_for_future_posts.
```

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+	 *
+	 * @return array[] Test parameters
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/post/wpUpdatePost.php]:
```suggestion
	 * @return array[] Test parameters.
```

**Comment by aaronjorbin** [comment]
> Finally, given that here is a good opportunity to open a specific file to cover wp_update_posts I've brought all the tests covering this function and merged them all into this file for Unit Tests organizational purposes.

This makes it harder to review this patch since it's not easy to see what the new tests are exactly. I think that should happen in a separate ticket.

**Comment by SirLouen** [comment]
> > Finally, given that here is a good opportunity to open a specific file to cover wp_update_posts I've brought all the tests covering this function and merged them all into this file for Unit Tests organizational purposes.
> 
> This makes it harder to review this patch since it's not easy to see what the new tests are exactly. I think that should happen in a separate ticket.

@aaronjorbin there are only 3 functions merged.
`test_update_invalid_post_id`
`test_update_post_should_only_modify_post_tags_if_different_tags_input_was_provided`
and
`test_wp_update_post_with_content_filtering`

They are literally copy/paste, so nothing to review there. Just for reunification purposes and I don't want them to be left to oblivion in the sea of forgotten tickets, specially if there are any future plans to move to PHPUnit higher version, its a free step forward.

There is only one function `test_update_post_preserves_date_for_future_posts` that targets this patch with a dataprovider. And its the last one.

By now I think it was finally reviewed by @peterwilsoncc 

The fix was simple and some of the examples in the dataprovider were a good idea of Peter. 

I think it's a straightforward merge at this point.

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+			)
+		);
+	}
+
+	/**
+	 * Test updating a post with an invalid ID.
+	 *
+	 * @ticket 23474
+	 */
+	public function test_update_invalid_post_id() {
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php:104]: Relocated test.

I agree with @aaronjorbin that relocating the tests makes this difficult to review. If it's to be done, then the assertions will need to be fixed to include messages.

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+	/**
+	 * Test ensuring that wp_update_post() does not unintentionally modify post tags
+	 * if the post has several tags with the same name but different slugs.
+	 *
+	 * Tags should only be modified if 'tags_input' parameter was explicitly provided,
+	 * and is different from the existing tags.
+	 *
+	 * @ticket 45121
+	 */
+	public function test_update_post_should_only_modify_post_tags_if_different_tags_input_was_provided() {
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php:124]: Relocated tests, as above re messages.

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+		$tags = wp_get_post_tags( $post->ID, array( 'fields' => 'ids' ) );
+		$this->assertSameSets( array( $tag_2['term_id'], $tag_3['term_id'] ), $tags );
+	}
+
+	/**
+	 * Test the wp_update_post() filters post content when 'post_status' is 'draft'.
+	 *
+	 * @ticket 22944
+	 */
+	public function test_wp_update_post_with_content_filtering() {
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php:161]: Relocated test, as above re messages.

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+			'ID'          => $post_id,
+			'post_status' => 'future',
+			'post_date'   => $future_date,
+		);
+
+		wp_update_post( $update_data );
+		$updated_post = get_post( $post_id );
+
+		$this->assertSame( $future_date, $updated_post->post_date );
+		$this->assertSame( $expected_status, $updated_post->post_status );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php]: These will need messages.

File: tests/phpunit/tests/post/wpUpdatePost.php [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+			),
+			'draft to future with 1 day more'   => array(
+				'initial_status'  => 'draft',
+				'time_offset'     => '+1 day',
+				'expected_status' => 'future',
+			),
+			'publish to future with 1 day more' => array(
+				'initial_status'  => 'publish',
+				'time_offset'     => '+1 day',
+				'expected_status' => 'publish',
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/wpUpdatePost.php]: I don't think this is correct.

On trunk:

1. Create and publish a post
2. Change the time to tomorrow
3. Change the status to scheduled
4. Click save
5. Observe in the database that the new status is `future`

**Review by peterwilsoncc — COMMENTED** [review]
I've added some notes inline.

The most important note is that the `publish to future with 1 day more` is expecting the status `publish` but the current behaviour on trunk is that the post will have the status `future`.

The current behaviour seems correct to me and should be retained.

**Comment by peterwilsoncc** [comment]
I'm unclear on the purpose of https://github.com/WordPress/wordpress-develop/pull/9699/commits/af483f2b158d206f28b7fb3c47bb6a033e761b66, when calling `wp_update|insert_post()` it's far more common to use the `post_date` argument rather than the `post_date_gmt` argument so it's of higher priority to test the former but both should probably be tested.

If the test is modified to compare ensure the timestamps for both date and date_gmt increases/decreases as appropriate, I'm seeing failures.

```php
public function test_update_post_preserves_date_for_future_posts( $initial_status, $time_offset, $expected_status ) {

	$post_id = self::factory()->post->create(
		array(
			'post_status' => $initial_status,
		)
	);

	$initial_date_timestamp = strtotime( get_post( $post_id )->post_date );
	$initial_date_gmt_timestamp = strtotime( get_post( $post_id )->post_date_gmt );

	$future_date = gmdate( 'Y-m-d H:i:s', strtotime( $time_offset ) );
	$update_data = array(
		'ID'            => $post_id,
		'post_status'   => 'future',
		'post_date_gmt' => $future_date,
	);

	wp_update_post( $update_data );
	$updated_post = get_post( $post_id );

	$updated_date_timestamp = strtotime( get_post( $post_id )->post_date );
	$updated_date_gmt_timestamp = strtotime( get_post( $post_id )->post_date_gmt );

	if ( substr( $time_offset, 0, 1 ) === '+' ) {
		$this->assertGreaterThan( $initial_date_timestamp, $updated_date_timestamp, 'Post date should be in the future compared to initial date.' );
		$this->assertGreaterThan( $initial_date_gmt_timestamp, $updated_date_gmt_timestamp, 'Post date GMT should be in the future compared to initial date GMT.' );
	} else {
		$this->assertLessThan( $initial_date_timestamp, $updated_date_timestamp, 'Post date should be in the past compared to initial date.' );
		$this->assertLessThan( $initial_date_gmt_timestamp, $updated_date_gmt_timestamp, 'Post date GMT should be in the past compared to initial date GMT.' );
	}

	$this->assertSame( $future_date, $updated_post->post_date_gmt, 'Post date should be updated accordingly.' );
	$this->assertSame( $expected_status, $updated_post->post_status, "Post status should be '{$expected_status}' after update." );
}
```

Result
```
$ phpunit --filter test_update_post_preserves_date_for_future_posts
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
Not running ajax tests. To execute these, use --group ajax.
Not running ms-files tests. To execute these, use --group ms-files.
Not running external-http tests. To execute these, use --group external-http.
PHPUnit 9.6.29 by Sebastian Bergmann and contributors.

Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!

FFFFFF                                                              6 / 6 (100%)

Time: 00:00.835, Memory: 219.89 MB

There were 6 failures:

1) Tests_Post_WpUpdatePost::test_update_post_preserves_date_for_future_posts with data set "pending to future with 1 day more" ('pending', '+1 day', 'future')
Post date should be in the future compared to initial date.
Failed asserting that 1762897192 is greater than 1762897192.

/vagrant/wordpress-develop/tests/phpunit/tests/post/wpUpdatePost.php:224

2) Tests_Post_WpUpdatePost::test_update_post_preserves_date_for_future_posts with data set "draft to future with 1 day more" ('draft', '+1 day', 'future')
Post date should be in the future compared to initial date.
Failed asserting that 1762897192 is greater than 1762897192.

/vagrant/wordpress-develop/tests/phpunit/tests/post/wpUpdatePost.php:224

3) Tests_Post_WpUpdatePost::test_update_post_preserves_date_for_future_posts with data set "publish to future with 1 day more" ('publish', '+1 day', 'future')
Post date should be in the future compared to initial date.
Failed asserting that 1762897192 is greater than 1762897192.

/vagrant/wordpress-develop/tests/phpunit/tests/post/wpUpdatePost.php:224

4) Tests_Post_WpUpdatePost::test_update_post_preserves_date_for_future_posts with data set "draft to future with 1 day less" ('draft', '-1 day', 'publish')
Post date should be in the past compared to initial date.
Failed asserting that 1762897192 is less than 1762897192.

/vagrant/wordpress-develop/tests/phpunit/tests/post/wpUpdatePost.php:227

5) Tests_Post_WpUpdatePost::test_update_post_preserves_date_for_future_posts with data set "pending to future with 1 day less" ('pending', '-1 day', 'publish')
Post date should be in the past compared to initial date.
Failed asserting that 1762897192 is less than 1762897192.

/vagrant/wordpress-develop/tests/phpunit/tests/post/wpUpdatePost.php:227

6) Tests_Post_WpUpdatePost::test_update_post_preserves_date_for_future_posts with data set "publish to future with 1 day less" ('publish', '-1 day', 'publish')
Post date should be in the past compared to initial date.
Failed asserting that 1762897192 is less than 1762897192.

/vagrant/wordpress-develop/tests/phpunit/tests/post/wpUpdatePost.php:227

FAILURES!
Tests: 6, Assertions: 6, Failures: 6.
```

**Comment by SirLouen** [comment]
Ok, I have updated with those tests. I will give it a second check to see what's going on now.
